### PR TITLE
Stricter validation for min/max values for whole numbers

### PR DESCRIPTION
**Pull request by fred84**
Currently we can overcame min/max validation by passing number as string or directly from plugin so its value will be rounded. Example:

```bash
curl -XPUT 'localhost:9200/twitter?pretty' -H 'Content-Type: application/json' -d'
{
  "mappings": {
    "tweet": {
      "properties": {
        "size": {
          "type": "long"
        }
      }
    }
  }
}
'

curl -XPUT 'localhost:9200/twitter/tweet/1' -H 'Content-Type: application/json' -d'{ "size":  9223372036854775806 }'
# long max value
curl -XPUT 'localhost:9200/twitter/tweet/2' -H 'Content-Type: application/json' -d'{ "size":  9223372036854775807 }'
# wrapped number > long max value
curl -XPUT 'localhost:9200/twitter/tweet/3' -H 'Content-Type: application/json' -d'{ "size":  "9223372036854775808" }'

curl -XGET 'localhost:9200/twitter/tweet/_search?pretty' -H 'Content-Type: application/json' -d'
{
    "query": {
        "match" : {
            "size" : 9223372036854775807
        }
    }
}
'

{
  #...
  "hits" : {
    "total" : 2,
    "max_score" : 1.0,
    "hits" : [
      {
        "_index" : "twitter",
        "_type" : "tweet",
        "_id" : "2",
        "_score" : 1.0,
        "_source" : {
          "size" : 9223372036854775807
        }
      },
      {
        "_index" : "twitter",
        "_type" : "tweet",
        "_id" : "3",
        "_score" : 1.0,
        "_source" : {
          "size" : "9223372036854775808"
        }
      }
    ]
  }
}
```

@colings86 @jpountz Please take a look.

### CODE REVIEW

**Comment by jpountz** [comment]
I'm wondering whether the fix should be upstream in jackson?

**Comment by fred84** [comment]
@jpountz There are 2 overloaded NumberType.parse methods. I agree that fix for ```parse(XContentParser parser, boolean coerce)``` fix should be in jackson, but for ```java parse(Object value, boolean coerce)``` I think it should be in elastic.

**Comment by jpountz** [comment]
Maybe I'm missing something but the latter seems to already perform bound checks?

**Comment by fred84** [comment]
@jpountz Incoming value in ```parse(Object value, boolean coerce)``` is converted to double before comparison against min/max value. In case of NumberType.LONG there is not enough precision for double near min/max value. 
Here is example demonstrating this: 
```assertFalse(Double.parseDouble("9223372036854776832") > Long.MAX_VALUE);```

**Comment by jpountz** [comment]
I am good with improving validation in that case.

**Comment by fred84** [comment]
I'll update PR. I think same validation for long should be added to ```AbstractXContentParser.public long longValue(boolean coerce)``` when handling string values.

**Comment by fred84** [comment]
@jpountz PR updated. I looked at Jackson's parser.getLongValue() - it validates min/max values correctly. Double to short/int/long conversion without proper validation was done in AbstractXContentParser.

File: core/src/main/java/org/elasticsearch/common/xcontent/support/AbstractXContentParser.java [hunk 1]
```diff
@@ -133,7 +138,14 @@ public short shortValue(boolean coerce) throws IOException {
...
             checkCoerceString(coerce, Short.class);
-            return (short) Double.parseDouble(text());
+
+            double doubleValue = Double.parseDouble(text());
+
+            if (doubleValue < Short.MIN_VALUE || doubleValue > Short.MAX_VALUE) {
+                throw new IllegalArgumentException("Value [" + text() + "] is out of range for a short");
+            }
+
+            return (short) doubleValue;
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/common/xcontent/support/AbstractXContentParser.java]: would it work if we only had `checkCoerceString` within this block? From what I read, `doShortValue` already has the ability to parse strings.

File: core/src/main/java/org/elasticsearch/common/xcontent/support/AbstractXContentParser.java [hunk 2]
```diff
@@ -176,14 +194,31 @@ public long longValue(boolean coerce) throws IOException {
...
-                return (long) Double.parseDouble(stringValue);
+                return preciseLongValue(stringValue, coerce);
             }
         }
         long result = doLongValue();
         ensureNumberConversion(coerce, result, Long.class);
         return result;
     }
 
+    public static long preciseLongValue(String stringValue, boolean coerce) {
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/common/xcontent/support/AbstractXContentParser.java]: if you want to use it from different classes, please move it to the `o.e.common.Numbers` class

**Review by jpountz — COMMENTED** [review]
I left some questions/comments but I think this is going in the right direction.

File: core/src/main/java/org/elasticsearch/common/xcontent/support/AbstractXContentParser.java (hunk 1 above)
Comment (fred84) [inline core/src/main/java/org/elasticsearch/common/xcontent/support/AbstractXContentParser.java]: `doShortValue` calls `parser.getShortValue()` which throws exception on non-numeric json values. Even if it is number in wrapped in string:
```java
    public void testGetShortValue() throws Exception {
        JsonFactory factory = new JsonFactory();

        JsonParser parser = factory.createParser("123");
        parser.nextToken();
        assertEquals(123, parser.getShortValue());

        final JsonParser parser2 = factory.createParser("\"123\"");
        parser2.nextToken();
        JsonParseException e = expectThrows(JsonParseException.class, () -> parser2.getShortValue());
        assertEquals("123", parser2.getValueAsString());
    }
```

File: core/src/main/java/org/elasticsearch/common/xcontent/support/AbstractXContentParser.java (hunk 2 above)
Comment (fred84) [inline core/src/main/java/org/elasticsearch/common/xcontent/support/AbstractXContentParser.java]: Ok

**Comment by fred84** [comment]
@jpountz PR updated. Also, when #25861 will be solved, we can futher simplify `NumericType.LONG.parse` to 
```java
Long parse(Object value, boolean coerce) {
    if (value instanceof Number) {
        return Numbers.toExactLong(value);
    }
    String stringValue = (value instanceof BytesRef) ? ((BytesRef) value).utf8ToString() : value.toString();
    return Numbers.toExactLong(stringValue);
}
```

**Review by jpountz — APPROVED** [review]
Thanks, it looks good to me.

**Comment by jpountz** [comment]
@elasticmachine please test it

**Comment by fred84** [comment]
@jpountz PR updated

**Comment by jpountz** [comment]
@elasticmachine please test it

**Pull request merged**